### PR TITLE
Media Type Multiplatform.

### PR DESCRIPTION
**Pull request by yschimke**
Deprecated members and Charset specific functionality have
been left out of common.

Tests have been split up into common, java specific and js specific.

It also introduces some modules, but I can revert if it's premature.

### CODE REVIEW

File: okhttp/src/commonTest/kotlin/okhttp3/MediaTypeTest.kt [hunk 1]
```diff
@@ -1,6 +1,5 @@
 /*
- * Copyright (C) 2013 Square, Inc.
- * Copyright (C) 2011 The Guava Authors
+ * Copyright (C) 2022 Square, Inc.
```
Comment (yschimke) [inline okhttp/src/commonTest/kotlin/okhttp3/MediaTypeTest.kt:2]: Is this copyright still warranted? cc @swankjesse

Comment (oldergod) [reply okhttp/src/commonTest/kotlin/okhttp3/MediaTypeTest.kt:2]: yes!

**Review by maldito-dot — COMMENTED** [review]
@maldito-dot

File: okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      "UTF-8", noCharset.charset(StandardCharsets.UTF_8)!!.name()
+    )
+    assertEquals(
+      "US-ASCII", noCharset.charset(StandardCharsets.US_ASCII)!!.name()
+    )
+    val charset = parse("text/plain; charset=iso-8859-1")
+    assertEquals(
+      "ISO-8859-1", charset.charset(StandardCharsets.UTF_8)!!.name()
+    )
+    assertEquals(
```
Comment (Goooler) [inline okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt:51]: We can replace `StandardCharsets` with `kotlin.text.Charsets` here?

File: okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      "ISO-8859-1", charset.charset(StandardCharsets.US_ASCII)!!.name()
```
Comment (Goooler) [inline okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt:52]: We can replace `StandardCharsets` with `kotlin.text.Charsets` here?

Comment (yschimke) [reply okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt:52]: Yep - it doesn't help much, these classes are JVM+withJava(), and most of the rest of the codebase still uses StandardCharsets.  If we change, we should change in one go?

Comment (Goooler) [reply okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt:52]: In #6986 do the rest.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -72,7 +72,32 @@ jobs:
...
-        run: ./gradlew test -Dtest.java.version=11
+        run: ./gradlew jvmTest -Dtest.java.version=11
+
+      - name: Publish Test Report
+        if: success() || failure()
+        uses: mikepenz/action-junit-report@v2
+        with:
+          report_paths: '**/build/test-results/test/TEST-*.xml'
+
+  testjs:
```
Comment (swankjesse) [inline .github/workflows/build.yml:83]: NICE

File: okhttp/src/commonMain/kotlin/okhttp3/MediaType.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  val type: String
+
+  /**
+   * Returns a specific media subtype, such as "plain" or "png", "mpeg", "mp4" or "xml".
+   */
+  val subtype: String
+
+  /** Alternating parameter names with their values, like `["charset", "utf-8"]`. */
+  internal val parameterNamesAndValues: Array<String>
```
Comment (swankjesse) [inline okhttp/src/commonMain/kotlin/okhttp3/MediaType.kt:43]: private?

File: okhttp/src/commonMain/kotlin/okhttp3/MediaType.kt [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * [rfc_2045]: http://tools.ietf.org/html/rfc2045
+ */
+expect class MediaType internal constructor(
+  mediaType: String,
+  type: String,
+  subtype: String,
+  parameterNamesAndValues: Array<String>,
+) {
+  internal val mediaType: String
```
Comment (swankjesse) [inline okhttp/src/commonMain/kotlin/okhttp3/MediaType.kt:30]: private?

File: okhttp/src/commonMain/kotlin/okhttp3/MediaType.kt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  fun parameter(name: String): String?
+
+  /**
+   * Returns the encoded media type, like "text/plain; charset=utf-8", appropriate for use in a
+   * Content-Type header.
+   */
+  override fun toString(): String
+
+  override fun equals(other: Any?): Boolean
```
Comment (swankjesse) [inline okhttp/src/commonMain/kotlin/okhttp3/MediaType.kt:57]: I don’t think these overrides are necessary? The `toString()` is good ’cause it specifies behavior.

File: okhttp/src/commonMain/kotlin/okhttp3/internal/-MediaTypeCommon.kt [hunk 8]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
```
Comment (swankjesse) [inline okhttp/src/commonMain/kotlin/okhttp3/internal/-MediaTypeCommon.kt:2]: nit: regular copyright statement with the newlines?

File: okhttp/src/commonTest/kotlin/okhttp3/MediaTypeTest.kt [hunk 9]
```diff
@@ -31,6 +28,8 @@ import okhttp3.internal.platform.Platform.Companion.isAndroid
  * MediaTypeTest.
  */
 open class MediaTypeTest {
+  open fun MediaType.charsetName(): String? = parameter("charset")?.uppercase()
```
Comment (swankjesse) [inline okhttp/src/commonTest/kotlin/okhttp3/MediaTypeTest.kt:31]: Good call

File: okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+import java.nio.charset.StandardCharsets
+import java.util.*
+import kotlin.test.assertEquals
+import kotlin.test.assertNull
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.internal.platform.Platform.Companion.isAndroid
+import org.junit.jupiter.api.Test
+
+class MediaTypeJavaTest: MediaTypeGetTest() {
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt:26]: Nit: Wanna use the suffix `Jvm` for JVM-specific Kotlin things, and `Java` for Java-language things?

File: okhttp/src/nonJvmMain/kotlin/okhttp3/MediaType.kt [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+package okhttp3
+
+import okhttp3.internal.commonEquals
+import okhttp3.internal.commonHashCode
+import okhttp3.internal.commonParameter
+import okhttp3.internal.commonToMediaType
+import okhttp3.internal.commonToMediaTypeOrNull
+import okhttp3.internal.commonToString
+
+actual class MediaType internal actual constructor(
```
Comment (swankjesse) [inline okhttp/src/nonJvmMain/kotlin/okhttp3/MediaType.kt:25]: Love it

**Review by swankjesse — APPROVED** [review]

File: okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt (hunk 10 above)
Comment (yschimke) [reply okhttp/src/jvmTest/java/okhttp3/MediaTypeJavaTest.kt]: Renamed, although I probably would have called Charset related functionality Java.  But assume Java means language.

JVM = Kotlin JVM specific
Java = Java language

**Pull request merged**